File: frontend/src/types/role.ts
export interface Role {
  id: string;
  name: string;
  description: string | null;
  permissions: Record<string, string>;
  is_system: boolean;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface RoleCreate {
  name: string;
  description?: string;
  permissions?: Record<string, string>;
}

export interface RoleUpdate {
  name?: string;
  description?: string;
  permissions?: Record<string, string>;
  is_active?: boolean;
}

File: backend/app/schemas/role.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class RoleCreate(BaseModel):
    name: str
    description: str | None = None
    permissions: dict[str, str] = {}  # {resource: level}, default to empty dict


class RoleUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    permissions: dict[str, str] | None = None
    is_active: bool | None = None


class RoleRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    description: str | None
    permissions: dict[str, str]
    is_system: bool
    is_active: bool
    created_at: datetime
    updated_at: datetime
